query-node/substrate-query-framework/index-builder/src/db/db.ts

```diff
@@ -1,54 +1,32 @@
-import { FindOneOptions, DeepPartial, EntityManager } from 'typeorm';
+import { FindOneOptions, DeepPartial } from 'typeorm';
 
 import { SubstrateEvent } from '..';
-import * as helper from './helper';
 
 /**
- * Database access object based on typeorm. Use typeorm transactional entity manager to perform get/save/remove
- * database operations.
- *
- * @constructor(event: QueryEvent, manager: EntityManager)
- *
+ * Database interaction interface.
```

Comment:
So here we  restrict the clients (e.g. the mapping functions) of this interface to only CRUD operations, right? It would be impossible to run custom queries, is it by design?

Reply:
That's right, only CRUD operations. I think we don't want the users to run custom queries. For example if they use typeorm `queryRunner` directly we will not be able to track the last processed event.

Reply:
OK, makes sense

Reply:
I think we should not allow the user to make custom `write` operations, but arbitrary `read` queries would be fine.

Its not very important, but would it be possible to offer a way to only do the latter, not the former?

Reply:
`typeorm` provides [`SelectQueryBuilder`](https://typeorm.io/#/select-query-builder). I tried it but it is not `read-only` safe. It means one can call `insert` on `select`.

```ts
await getConnection()
    .createQueryBuilder()
    .select('membership')
    .from(Membership, 'membership')
    .insert() // It works
    .into(Membership)
    .values([{ memberId: 234, accountId: '5FdyqzHPZe71EzDU3pHKBWEPChbUitfoToUgrDtnWoojFtic' }])
    .execute();
```

Right now you can perform `join`, `relations`, `order` with `db.get()`. I think if we add `getMany()` that would be enough

Reply:
Ok great, that is quite good actually.